ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobProvider.java

```diff
@@ -84,13 +84,24 @@ public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundEx
       return null;
     }
     ParcelFileDescriptor readSide = pipe[0];
-    ParcelFileDescriptor writeSide = pipe[1];
-
-    try (OutputStream outputStream = new ParcelFileDescriptor.AutoCloseOutputStream(writeSide)) {
-      outputStream.write(data);
-    } catch (IOException exception) {
-      return null;
-    }
+    final ParcelFileDescriptor writeSide = pipe[1];
+
+    Thread writer =
+        new Thread() {
+          public void run() {
+            try (OutputStream outputStream =
+                new ParcelFileDescriptor.AutoCloseOutputStream(writeSide)) {
+              // If the blob data is larger than pipe capacity (64 KB), a synchronous write call
+              // would fill up the whole buffer and block until the bytes are read (i.e. never).
+              // Writing from a separate thread allows us to return the read side descriptor
+              // immediately so that the reader can start reading.
+              outputStream.write(data);
+            } catch (IOException exception) {
+              // no-op
```

Comment:
What happens in this situation? Should we somehow close or notify with failure?

Previously, the parent method would return `null` instead of `readSide`.

Reply:
Since the new implementation returns a `ParcelFileDescriptor` immediately, i.e. before calling `write` (which possibly could fail later on), we cannot return `null` as previously.

`AutoCloseOutputStream` ensures that the write side descriptor always gets closed, no matter if the write was successful or not. In case of an I/O error, the write side descriptor gets closed, and reading from a broken pipe will also result in an `IOException` on the reader side.

According to [Android documentation](https://developer.android.com/reference/android/content/ContentProvider#openFile(android.net.Uri,%20java.lang.String)), the reader is responsible for closing the read side descriptor:

> The returned ParcelFileDescriptor is owned by the caller, so it is their responsibility to close it when done.

So it should be safe to return `ParcelFileDescriptor` instead of `null`, since the reader is responsible for closing it.

The opposite case is when the reader closes the read side descriptor on purpose, for example when unmounting an image component before it was fully loaded (i.e. partial read). In such case, the write call fails and the write side descriptor gets closed as well, so there is no resource leak.

Reply:
Amazing! Thank you for the detailed explanation.